hybrid/ddc_null.c

```diff
@@ -15,6 +16,9 @@
 int main()
 {
 	pp_cap(cheri_ddc_get());
+	// The DDC is empty and we ensure it is
```

Comment:
What does "The DDC is empty" mean? I assume it means "the DDC is NULL"? And I guess that can only happen if this file is compiled in purecap mode?

Reply:
Well, if NULL it would crash. In hybrid-mode when we haven't put anything inside it, it's address is all zeros. It contains a valid tag and address with all zeros.

Reply:
"All zeros" is NULL on the platforms we care about.

> And I guess that can only happen if this file is compiled in purecap mode?

Can I ping you on this bit too please.

Reply:
Here we are in hybrid mode, so we are not talking about purecap. If what you said were true, the assertions would fail. This example is hybrid mode only.

Reply:
OK let's either a) add a compile time `#ifdef`-esque check that we're not in purecap mode (preferably) or b) at least document that this is hybrid only. Can we do that for all the examples please?

Reply:
All the examples here were developed for hybrid mode. If you mean the whole repo, of course, we can do it. Here were are experimenting in hybrid mode, unless I have missed some pieces along the way...

Reply:
And I'll document this better.

Reply:
For the files in this PR, let's add compile-time checks (if possible -- which I think it is), as they're the best form of documentation! It doesn't hurt to add human-friendly documentation to the same effect too, of course!

Reply:
> "All zeros" is NULL on the platforms we care about.
> 
> > And I guess that can only happen if this file is compiled in purecap mode?
> 
> Can I ping you on this bit too please.

This is how the DDC looks when compiling in hybrid mode and printing is content at the beginning of the program:
```
Capability: 0x0 [rwRW,0x0-0x1000000000000]
Tag: 1, Perms: 3717d, Type: 0, Address: 0000, Base: 0000, End: 1000000000000, Flags: 0, Length: 1000000000000, Offset: 0000
```

As you can see, the capability is valid but it's address is zero. When we put a `NULL` into it, the capability will become *invalid*. You wanted to clarify this point further, so I thought this would have been a good place.

Reply:
This isn't a NULL capability: it has a NULL address but notice that the `End` is != 0. In other words, it's more-or-less the DDC I would have expected: its bounds cover what I assume is the entirety of the virtual address?

Reply:
> This isn't a NULL capability: it has a NULL address but notice that the `End` is != 0. In other words, it's more-or-less the DDC I would have expected: its bounds cover what I assume is the entirety of the virtual address?

I had referred to this as "EMPTY" but, of course, is a wrong naming. I *think* you are right regarding its bounds.

Reply:
Anyway, let's please put the `#ifdef`-esque checks that mean that we're only compiling this in hybrid mode. It should be easy to do (I think!) and will help remind our future selves what's going on in the files that result from this PR.

Reply:
> Anyway, let's please put the `#ifdef`-esque checks that mean that we're only compiling this in hybrid mode. It should be easy to do (I think!) and will help remind our future selves what's going on in the files that result from this PR.

OK.

Reply:
> Anyway, let's please put the `#ifdef`-esque checks that mean that we're only compiling this in hybrid mode. It should be easy to do (I think!) and will help remind our future selves what's going on in the files that result from this PR.

Are you thinking about custom macro definition to check a capability or something different? Because I can't find anything in the `cheriintrin.h`...

Reply:
Something like https://github.com/capablevms/cheri_misidioms/blob/master/code/narrow.c#L7 (but not purecap!).

Reply:
> Something like https://github.com/capablevms/cheri_misidioms/blob/master/code/narrow.c#L7 (but not purecap!).

It seems the only way (they also use only the purecap check in all the projects). There is no equivalent like "*_HYBRID". So checking we are in purecap but we should not be (since this PR is hybrid only), I think should be sufficient.

Reply:
Fix in 28ade2d.

Reply:
Technically this will compile on any normal architecture. I *think* we can get "CHERI hybrid mode" with something like:

```c
#if !defined(__CHERI_CAPABILITY_WIDTH__)
#error "Only works on CHERI"
#elif defined(__CHERI_PURE_CAPABILITY__)
#error "Only works on CHERI hybrid mode"
#endif
```

We could squish that down to:

```c
#if defined(__CHERI_CAPABILITY_WIDTH__) || !defined(__CHERI_PURE_CAPABILITY__)
#error "Only works on CHERI hybrid mode"
#endif
```

I think.

Reply:
You are quite right.

Reply:
> Technically this will compile on any normal architecture. I _think_ we can get "CHERI hybrid mode" with something like:
> 
> ```c
> #if !defined(__CHERI_CAPABILITY_WIDTH__)
> #error "Only works on CHERI"
> #elif defined(__CHERI_PURE_CAPABILITY__)
> #error "Only works on CHERI hybrid mode"
> #endif
> ```
> 
> We could squish that down to:
> 
> ```c
> #if defined(__CHERI_CAPABILITY_WIDTH__) || !defined(__CHERI_PURE_CAPABILITY__)
> #error "Only works on CHERI hybrid mode"
> #endif
> ```
> 
> I think.

It seems `__CHERI_CAPABILITY_WIDTH__` (128) is defined even on *hybrid mode*.

Reply:
> It seems __CHERI_CAPABILITY_WIDTH__ (128) is defined even on hybrid mode.

Oops, I've got my logic the wrong way around! It should be:

```c
#if !defined(__CHERI_CAPABILITY_WIDTH__) || defined(__CHERI_PURE_CAPABILITY__)
#error "Only works on CHERI hybrid mode"
#endif
```

Reply:
Addressed in 108c26a.